### PR TITLE
ELPP-2252 KVS Cache for transformer

### PR DESCRIPTION
**Pull request by stephenwf**
@giorgiosironi  @thewilkybarkid

### CODE REVIEW

**Comment by giorgiosironi** [comment]
I understand the need of storing subjects data, but given a single SQS queue how do we avoid doing this for all types that come out from its messages? Is there some routing done from the client code of the library? Will reread it tomorrow morning.

**Comment by stephenwf** [comment]
yes so here in search:
https://github.com/elifesciences/search/blob/develop/src/Search/Gearman/Command/QueueWatchCommand.php#L49-L50

We have 2 transforms, one for `QueueItem` to gearman task and another `QueueItem` to ApiSDK item. We can add to those lines: 

```php
if($item->getType() === 'subject') {
    $entity = $this->cachedTransformer->transform($item);
}
```

Or something along those lines.

Alternatively I could add a callable "shouldCache" to the constructor that will dictate when to cache.

**Comment by stephenwf** [comment]
Okay I've added the `$shouldCache` property, I think that will work

File: src/Queue/CachedTransformer.php [hunk 1]
```diff
@@ -15,18 +15,37 @@ class CachedTransformer implements QueueItemTransformer
...
         $this->sdk = $sdk;
         $this->cache = $cache;
         $this->logger = $logger;
         $this->lifetime = $lifetime;
+        $this->shouldCache = $shouldCache;
+    }
+
+    private function shouldCacheEntity(string $type, string $id)
+    {
+        if ($shouldCache = $this->shouldCache) {
```
Comment (thewilkybarkid) [inline src/Queue/CachedTransformer.php]: The `if` is redundant isn't it?

File: src/Queue/CachedTransformer.php [hunk 2]
```diff
@@ -15,18 +15,37 @@ class CachedTransformer implements QueueItemTransformer
...
+            return $shouldCache($type, $id);
+        }
+
+        return true;
     }
 
     public function get($id, $type)
     {
+        if (!$this->shouldCacheEntity($id, $type)) {
```
Comment (thewilkybarkid) [inline src/Queue/CachedTransformer.php]: This should probably come last (what if the cache entry has disappeared?).

**Comment by giorgiosironi** [comment]
I still do not understand the integration with the client project. Using a cache inside `queue:watch` does not seem a good idea because `queue:watch` is notified of changes, hence if it then caches the GET requests it performs a change can be lost due to a stale cache?

I thought this kind of cache should be written and read by the HTTP processes that returns responses to the API calls (to fill in `subjects`-related fields in the response), not by the long-running processes?

File: src/Queue/CachedTransformer.php (hunk 1 above)
Comment (stephenwf) [inline src/Queue/CachedTransformer.php]: Ah forgot to make the should cache optional in the constructor, avoids something like `function() { return true; }` in recommendations

**Comment by stephenwf** [comment]
@giorgiosironi  So the `queue:watch` will only use the `transform()` command like it does just now, when it calls that it will always 100% make a network request to get the item. If its been told to cache that item it will add it to the cache. 

Only the user-facing controller will use the `get()` method hitting the cache

**Comment by giorgiosironi** [comment]
But why tying up this cache with the Transformers and the queue at all, if we only need it on the serving side (which I assume does not use Transformers and these classes)? It can save some dozen calls to populate the cache from the transform() that can then be used in the get(), but it's a lot of complexity for little gain?

**Comment by stephenwf** [comment]
For recommendations we need a cache that will server 99% of the content imported, the calls to populate can be handled in the cache itself, for the initial import we can have cache that will batch the calls so its not trashing the kvs. The gain is simply cache invalidation, with this we can always ensure fresh content. (without Dos-ing journal 👀 )

File: src/Queue/CachedTransformer.php [hunk 3]
```diff
@@ -0,0 +1,108 @@
+<?php
+
+namespace eLife\Bus\Queue;
+
+use Doctrine\Common\Cache\Cache;
+use eLife\ApiSdk\ApiSdk;
+use eLife\ApiSdk\Model\Model;
+use Psr\Log\LoggerInterface;
+
```
Comment (giorgiosironi) [inline src/Queue/CachedTransformer.php]: could you add a docblock here that explains the difference between `get()` and `refresh()` and the use cases from which they are called?

Comment (giorgiosironi) [inline src/Queue/CachedTransformer.php]: want this to be clear to people reading it in months time

File: src/Queue/CachedTransformer.php [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    private function shouldCacheEntity(string $type, string $id)
+    {
+        if ($shouldCache = $this->shouldCache) {
+            return $shouldCache($type, $id);
+        }
+
+        return true;
+    }
+
+    public function get($id, $type)
```
Comment (giorgiosironi) [inline src/Queue/CachedTransformer.php]: an interface for this `get()` method would be useful as code in the serving phase of the service will not depend on a `CachedTransformer` but rather on a smaller interface

**Comment by stephenwf** [comment]
@giorgiosironi I've added an interface and changed up the classes that implement the interface. Its possible these two bits of functionality could be split out if that makes more sense in each implementation. 

Also described the 3 scenarios for the `get()` command, cache, no-cache and cache but stale.

File: src/Queue/CachedTransformer.php [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    private function getKeyFromQueueItem(QueueItem $item) : string
+    {
+        return static::getKey($item->getType(), $item->getId());
+    }
+
+    public static function getKey(string $type, string $id) : string
+    {
+        return sha1(sha1($id), sha1($type));
```
Comment (giorgiosironi) [inline src/Queue/CachedTransformer.php]: why not just `sha1("$type}:{$id}")` :thinking:

**Comment by giorgiosironi** [comment]
Segregated interfaces do not necessarily imply segregated implementations, good to go :+1:

File: src/Queue/CachedTransformer.php (hunk 5 above)
Comment (thewilkybarkid) [inline src/Queue/CachedTransformer.php]: Second argument of `sha1()` is a boolean too.

Comment (stephenwf) [inline src/Queue/CachedTransformer.php]: 👍  sorry, fixed

**Pull request merged**